File: experiments/adrianco/experiment-7/brazil/runs/language=rust_model=claude-opus-4-8-fast/rep1/src/mcp.rs
// =============================================================================
// Context: Brazilian Soccer MCP Server
// File:    src/mcp.rs
// Purpose: A minimal, dependency-light implementation of the Model Context
//          Protocol JSON-RPC 2.0 message layer over stdio. Handles the
//          `initialize` handshake, `tools/list`, `tools/call` and `ping`, and
//          turns tool output into MCP `content` blocks. Notifications (such as
//          `notifications/initialized`) yield no response.
//
//          Keeping the protocol logic separate from the data/query/tool layers
//          means the entire server can be driven in tests by feeding request
//          `Value`s to `handle_request`.
// =============================================================================

use serde_json::{json, Value};

use crate::data::Database;
use crate::tools;

/// MCP protocol revision advertised to clients.
pub const PROTOCOL_VERSION: &str = "2024-11-05";

/// Handle a single JSON-RPC request. Returns `Some(response)` for requests and
/// `None` for notifications (messages without an `id`).
pub fn handle_request(db: &Database, request: &Value) -> Option<Value> {
    let id = request.get("id").cloned();
    let method = request.get("method").and_then(|m| m.as_str()).unwrap_or("");
    let params = request.get("params").cloned().unwrap_or(Value::Null);

    // Notifications carry no id and never get a response.
    let id = id?;

    match method {
        "initialize" => Some(success(
            id,
            json!({
                "protocolVersion": PROTOCOL_VERSION,
                "capabilities": { "tools": {} },
                "serverInfo": {
                    "name": "brazilian-soccer-mcp",
                    "version": env!("CARGO_PKG_VERSION")
                },
                "instructions": "Query Brazilian soccer match and player data. Use list_competitions to see coverage, then search_matches / team_stats / head_to_head / search_players / competition_standings / competition_stats."
            }),
        )),
        "ping" => Some(success(id, json!({}))),
        "tools/list" => Some(success(id, json!({ "tools": tools::tool_definitions() }))),
        "tools/call" => Some(handle_tools_call(db, id, &params)),
        other => Some(error(id, -32601, &format!("Method not found: {other}"))),
    }
}

fn handle_tools_call(db: &Database, id: Value, params: &Value) -> Value {
    let name = match params.get("name").and_then(|n| n.as_str()) {
        Some(n) => n,
        None => return error(id, -32602, "Missing tool name"),
    };
    let args = params.get("arguments").cloned().unwrap_or_else(|| json!({}));
    match tools::call_tool(db, name, &args) {
        Ok(text) => success(
            id,
            json!({
                "content": [ { "type": "text", "text": text } ],
                "isError": false
            }),
        ),
        Err(message) => success(
            id,
            json!({
                "content": [ { "type": "text", "text": message } ],
                "isError": true
            }),
        ),
    }
}

fn success(id: Value, result: Value) -> Value {
    json!({ "jsonrpc": "2.0", "id": id, "result": result })
}

fn error(id: Value, code: i64, message: &str) -> Value {
    json!({ "jsonrpc": "2.0", "id": id, "error": { "code": code, "message": message } })
}

File: experiments/adrianco/experiment-46-opus5/brazil/runs/language=clojure_model=claude-opus-5_prompt=neutral/rep1/src/brazilian_soccer/mcp.clj
(ns brazilian-soccer.mcp
  "Model Context Protocol server: JSON-RPC 2.0 over newline delimited stdio.

  CONTEXT
  -------
  MCP clients (Claude Desktop, the MCP inspector, any LLM host) launch this
  process and speak JSON-RPC on its stdin/stdout, one JSON object per line.
  The handshake and the methods implemented here are:

      initialize                -> protocol version, capabilities, server info
      notifications/initialized -> acknowledged silently (no response)
      ping                      -> {}
      tools/list                -> the catalogue from brazilian-soccer.tools
      tools/call                -> {content:[{type:\"text\",...}], isError}

  Anything else gets a JSON-RPC -32601.  Tool failures are *not* JSON-RPC
  errors: MCP wants them reported as a result with isError true so the model
  can read the message and retry, which is how \"did you mean Atlético Mineiro
  or Athletico Paranaense?\" reaches the user.

  `handle-message` is a pure function of [db message] so the protocol can be
  tested without spawning a process; `serve` is the thin I/O loop."
  (:require [clojure.data.json :as json]
            [clojure.string :as str]
            [brazilian-soccer.tools :as tools]))

(def protocol-version "2024-11-05")
(def supported-protocol-versions #{"2024-11-05" "2025-03-26" "2025-06-18"})

(def server-info {"name" "brazilian-soccer" "version" "1.0.0"})

(def instructions
  (str "Knowledge graph of Brazilian football built from six Kaggle datasets: "
       "16,000+ matches (Brasileirão Série A/B/C 2003-2023, Copa do Brasil 2012-2023, "
       "Copa Libertadores 2013-2022) and the FIFA 19 player database. "
       "Call dataset_info first if you are unsure what is covered. Club names may be "
       "written in any spelling found in the data (\"Flamengo\", \"Flamengo-RJ\", "
       "\"Atletico Mineiro\", \"Atlético - MG\"). Standings and averages are calculated "
       "from the match results present in the data, not copied from an official table."))

(defn- ok [id result] {"jsonrpc" "2.0" "id" id "result" result})

(defn- rpc-error [id code message]
  {"jsonrpc" "2.0" "id" id "error" {"code" code "message" message}})

(defn- text-result [text error?]
  {"content" [{"type" "text" "text" text}] "isError" (boolean error?)})

(defn- initialize-result [params]
  (let [requested (get params "protocolVersion")]
    {"protocolVersion" (if (supported-protocol-versions requested) requested protocol-version)
     "capabilities" {"tools" {"listChanged" false}}
     "serverInfo" server-info
     "instructions" instructions}))

(defn handle-message
  "Handle one decoded JSON-RPC message.  Returns the response map, or nil for
  notifications (messages without an id), which must not be answered."
  [db message]
  (let [{:strs [id method params]} message
        notification? (not (contains? message "id"))]
    (cond
      (nil? method)
      (when-not notification? (rpc-error id -32600 "Invalid request: no method"))

      (= method "initialize") (ok id (initialize-result params))
      (= method "ping")       (ok id {})

      (str/starts-with? method "notifications/") nil

      (= method "tools/list")
      (ok id {"tools" (tools/tool-list-json)})

      (= method "tools/call")
      (let [tool-name (get params "name")
            args      (or (get params "arguments") {})]
        (try
          (ok id (text-result (tools/call-tool db tool-name args) false))
          (catch Exception e
            (ok id (text-result (str "Tool " tool-name " failed: "
                                     (or (.getMessage e) (str e)))
                                true)))))

      :else
      (when-not notification?
        (rpc-error id -32601 (str "Method not found: " method))))))

(defn handle-line
  "Decode a line, dispatch it and encode the response (nil when silent)."
  [db line]
  (let [message (try (json/read-str line)
                     (catch Exception e
                       {::parse-error (.getMessage e)}))]
    (cond
      (::parse-error message)
      (json/write-str (rpc-error nil -32700 (str "Parse error: " (::parse-error message))))

      (not (map? message))
      (json/write-str (rpc-error nil -32600 "Invalid request: expected a JSON object"))

      :else
      (some-> (handle-message db message) json/write-str))))

(defn serve
  "Read newline delimited JSON-RPC from `in`, write responses to `out`.
  Both streams are forced to UTF-8: club names carry accents and cedillas and
  the platform default encoding cannot be trusted."
  [db in out]
  (let [reader (java.io.BufferedReader. (java.io.InputStreamReader. in "UTF-8"))
        writer (java.io.PrintWriter. (java.io.OutputStreamWriter. out "UTF-8") true)]
    (doseq [line (line-seq reader)
            :when (not (str/blank? line))]
      (when-let [response (handle-line db line)]
        (.println writer response)
        (.flush writer)))))
